File: backend/app/schemas/pending_item_report.py
import uuid

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.pending_item import PendingItemCompliance, PendingItemState


class PendingItemReportPeriod(BaseModel):
    model_config = ConfigDict(extra="forbid")

    planned_from: date | None
    planned_to: date | None


class PendingItemReportFilters(BaseModel):
    model_config = ConfigDict(extra="forbid")

    category_id: uuid.UUID | None
    is_active: bool | None
    state: PendingItemState | None
    compliance: PendingItemCompliance | None


class PendingItemReportSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    total_count: int = Field(ge=0)
    active_count: int = Field(ge=0)
    inactive_count: int = Field(ge=0)
    no_iniciado_count: int = Field(ge=0)
    en_proceso_count: int = Field(ge=0)
    finalizado_count: int = Field(ge=0)


class PendingItemComplianceSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    en_plazo_count: int = Field(ge=0)
    atrasado_count: int = Field(ge=0)
    con_adelanto_count: int = Field(ge=0)
    a_tiempo_count: int = Field(ge=0)
    con_retraso_count: int = Field(ge=0)


class PendingItemDetailSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    average_atrasado_days: Decimal | None = Field(default=None, ge=0)
    average_con_adelanto_days: Decimal | None = Field(default=None, ge=0)
    average_con_retraso_days: Decimal | None = Field(default=None, ge=0)


class PendingItemCategoryReportRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    category_id: uuid.UUID
    category_name: str
    summary: PendingItemReportSummary
    compliance: PendingItemComplianceSummary


class PendingItemReportResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    period: PendingItemReportPeriod
    filters: PendingItemReportFilters
    summary: PendingItemReportSummary
    compliance: PendingItemComplianceSummary
    detail: PendingItemDetailSummary
    by_category: list[PendingItemCategoryReportRow]

File: frontend/src/types/pendingItemReport.ts
export type PendingItemState = "NO_INICIADO" | "EN_PROCESO" | "FINALIZADO";
export type PendingItemCompliance =
  | "EN_PLAZO"
  | "ATRASADO"
  | "CON_ADELANTO"
  | "A_TIEMPO"
  | "CON_RETRASO";

export interface PendingItemReportParams {
  planned_from?: string;
  planned_to?: string;
  category_id?: string;
  is_active?: boolean;
  state?: PendingItemState;
  compliance?: PendingItemCompliance;
}

export interface PendingItemReportSummary {
  total_count: number;
  active_count: number;
  inactive_count: number;
  no_iniciado_count: number;
  en_proceso_count: number;
  finalizado_count: number;
}

export interface PendingItemComplianceSummary {
  en_plazo_count: number;
  atrasado_count: number;
  con_adelanto_count: number;
  a_tiempo_count: number;
  con_retraso_count: number;
}

export interface PendingItemDetailSummary {
  average_atrasado_days: string | number | null;
  average_con_adelanto_days: string | number | null;
  average_con_retraso_days: string | number | null;
}

export interface PendingItemCategoryReportRow {
  category_id: string;
  category_name: string;
  summary: PendingItemReportSummary;
  compliance: PendingItemComplianceSummary;
}

export interface PendingItemReportResponse {
  period: { planned_from: string | null; planned_to: string | null };
  filters: {
    category_id: string | null;
    is_active: boolean | null;
    state: PendingItemState | null;
    compliance: PendingItemCompliance | null;
  };
  summary: PendingItemReportSummary;
  compliance: PendingItemComplianceSummary;
  detail: PendingItemDetailSummary;
  by_category: PendingItemCategoryReportRow[];
}
